File: 송하연/10월 4주차/퍼즐 게임 챌린지.java
import java.util.*;

class Solution {
    public int solution(int[] diffs, int[] times, long limit) {
        int answer = 0;
        
        int max = 0;
        for(int diff:diffs){
            max = Math.max(max,diff);
        }
        
        int low=1;
        int high=max;
        
        while(low<=high){
            int mid = (low+high)/2;
            if(solve(mid,diffs,times)<=limit) {
                answer= mid;
                high=mid-1;
            }
            else low=mid+1;
        }
    
        return answer;
    }
    
    long solve(int level, int[] diffs, int[] times){
        long totalTime = 0;
        totalTime+= times[0];
        
        for(int i=1; i<diffs.length; i++){  
            if(diffs[i] <= level) totalTime+= times[i];
            //diff > level : 현재 시간 + 이전 시간 * 실패 횟수(diff-level) + 현재시간
            else{
                totalTime+= (times[i]+times[i-1]) * (diffs[i]-level) + times[i];
            }
        }
        //System.out.println(totalTime);
        return totalTime; 
    }
}

File: 한승우/10월 5주차/퍼즐 게임 챌린지.cpp
#include <string>
#include <vector>

using namespace std;

typedef long long ll;

/*
diff > level
-> (diff - level) * (time_cur + time_prev) + time_cur //틀리고, 맞고
*/

ll cal(ll level, const vector<int>& diffs, const vector<int>& times){
    ll time = times[0];
    
    for(int i = 1; i < diffs.size(); i++){
        int time_prev = times[i - 1];
        int time_cur = times[i];
        
        if(diffs[i] <= level)
            time += time_cur;
        else 
            time += (diffs[i] - level) * (time_cur + time_prev) + time_cur;
    }
    
    return time;
}

int solution(vector<int> diffs, vector<int> times, ll limit) {
    ll st = 1, end = 1e15;
    
    while(st < end){
        ll mid = (st + end) / 2;
        
        if(cal(mid, diffs, times) <= limit) end = mid;
        else st = mid + 1;
    }
    
    return st;
}s